File: components/support/nimbus-fml/src/backends/kotlin/templates/macros.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

 {#- This file contains macros needed to generate code for the FML.

    It is the natural place to put commonalities between Object and Features,
    and rendering literals for Objects.
-#}

{% macro render_constructor() -%}
private constructor(
   private val _variables: Variables,
   private val _prefs: SharedPreferences? = null,
   private val _defaults: Defaults)
{% endmacro %}

{% macro render_class_body(inner) %}
{%- let prop_name = inner.name()|var_name %}
{%- let raw_name = inner.name() -%}
   {# The data class holds the default values that come from the manifest. They should completely
   specify all values needed for the  feature #}
   private data class Defaults({% for p in inner.props() %}
      {%- let t = p.typ() %}
      val {{p.name()|var_name}}: {{ t|defaults_type_label }}{% if !loop.last %},{% endif %}
   {%- endfor %}
   )

   {#- A constructor for application tests to use.  #}

   constructor(_variables: Variables = NullVariables.instance, _prefs: SharedPreferences? = null, {% for p in inner.props() %}
   {%- let t = p.typ() %}
      {{p.name()|var_name}}: {{ t|defaults_type_label }} = {{ t|literal(self, p.default(), "_variables.context") }}{% if !loop.last %},{% endif %}
   {%- endfor %}
   ) : this(
      _variables = _variables,
      _prefs = _prefs,
      _defaults = Defaults({% for p in inner.props() %}
      {%- let nm = p.name()|var_name %}{{ nm }} = {{ nm }}{% if !loop.last %}, {% endif %}
      {%- endfor %})
   )

   {# The property initializers #}
   {%- for p in inner.props() %}
   {%- let prop_kt = p.name()|var_name %}
   {%- let type_kt = p.typ()|type_label %}
   {%- let defaults = format!("_defaults.{}", prop_kt) %}
   {%- let getter = p.typ()|property(p.name(), "_variables", defaults) %}
  {{ p.doc()|comment("  ") }}
   {%- if p.supports_lazy() %}
   val {{ prop_kt }}: {{ type_kt }} by lazy {
      {{ getter }}
   }
   {%- else %}
   val {{ prop_kt }}: {{ type_kt }}
      get() {
         fun getter() = {{ getter }}
         return {% call prefs() %}?.let { _ ->
            getter()
         } ?: getter()
      }
   {%- endif %}
{% endfor %}

   {#- toJSON #}
   override fun toJSONObject(): JSONObject =
      JSONObject(
         mapOf(
            {%- for p in inner.props() %}
            {{ p.name()|quoted }} to {{ p.name()|var_name|to_json(p.typ()) }},
            {%- endfor %}
         )
      )
{% endmacro %}

{% macro prefs() %}this._prefs{% endmacro %}

File: components/support/nimbus-fml/src/backends/swift/templates/macros.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

 {#- This file contains macros needed to generate code for the FML.

    It is the natural place to put commonalities between Object and Features,
    and rendering literals for Objects.
-#}

{%- macro render_class(inner) %}
{%- let raw_name = inner.name() %}
{% let class_name = inner.name()|class_name -%}

{{ inner.doc()|comment("") }}
public class {{class_name}}: FMLObjectInterface {
    private let _variables: Variables
    private let _defaults: Defaults
    private let _prefs: UserDefaults?
    private init(variables: Variables = NilVariables.instance, prefs: UserDefaults? = nil, defaults: Defaults) {
        self._variables = variables
        self._defaults = defaults
        self._prefs = prefs
    }
    {# The struct holds the default values that come from the manifest. They should completely
    specify all values needed for the  feature #}
    struct Defaults {
        {%- for p in inner.props() %}
        {%- let t = p.typ() %}
        let {{p.name()|var_name}}: {{ t|defaults_type_label }}
    {%- endfor %}
    }

    {#- A constructor for application tests to use.  #}

    public convenience init(
        _ _variables: Variables = NilVariables.instance,
        _ _prefs: UserDefaults? = nil,
        {%- for p in inner.props() %}
        {%- let t = p.typ() %}
        {{p.name()|var_name}}: {{ t|defaults_type_label }} = {{ t|literal(self, p.default(), "") }}{% if !loop.last %},{% endif %}
    {%- endfor %}
    ) {
        self.init(variables: _variables, prefs: _prefs, defaults: Defaults({% for p in inner.props() %}
            {{p.name()|var_name}}: {{p.name()|var_name}}{% if !loop.last %},{% endif %}
        {%- endfor %}))
    }

{#- The property initializers #}
{# -#}
    {% for p in inner.props() %}
    {%- let prop_swift = p.name()|var_name %}
    {%- let type_swift = p.typ()|type_label %}
    {%- let defaults = format!("_defaults.{}", prop_swift) %}
    {%- let getter = p.typ()|property(p.name(), "self._variables", defaults) %}
    {{ p.doc()|comment("    ") }}
    {%- if !p.has_prefs() %}
    public lazy var {{ prop_swift }}: {{ type_swift }} = {
        {{ getter }}
    }()
    {%- else %}
    public var {{ prop_swift }}: {{ type_swift }} {
        {%- let prefs = "prefs" %}
        {%- let key = p.pref_key().unwrap() %}
        {#- Using `object(forKey:)` here as it returns an optional that just needs to be cast.
            `integer(forKey:)` returns zero, `bool(forKey:)` returns `false` if the key is not
            present.

            Now we only need to use the type label as part of the cast, we don't need to implement
            separate type specific preference getters.

            `has_prefs()` checks if the type can be got from UserDefaults.
            #}
        if let {{ prefs }} = {% call prefs() %},
            let {{ prop_swift }} = {{ prefs }}.object(forKey: {{ key|quoted }}) as? {{ type_swift }} {
            return {{ prop_swift }}
        }
        return {{ getter }}
    }
    {%- endif %}
    {% endfor %}
}

{%- endmacro %}}

{% macro prefs() %}self._prefs{% endmacro %}
